repository/kafka/operator/operator.yaml

```diff
@@ -34,38 +48,35 @@ tasks:
         - file: /tmp/tls.crt
           kind: Secret
           key: certificate
-  - name: user-workload
-    kind: Apply
+  # feature tasks
+  - name: service-monitor
+    kind: Toggle
     spec:
+      parameter: ADD_SERVICE_MONITOR
       resources:
-      - user-workload.yaml
-  - name: service
-    kind: Apply
+        - service-monitor.yaml
+  - name: external-access
+    kind: Toggle
     spec:
+      parameter: EXTERNAL_ADVERTISED_LISTENER
       resources:
-      - service.yaml
-      - service-monitor.yaml
-  - name: external-access
-    kind: Apply
+        - clusterrole.yaml
+        - clusterrolebinding.yaml
+        - external-service.yaml
+  - name: mirrormaker
+    kind: Toggle
     spec:
+      parameter: MIRROR_MAKER_ENABLED
       resources:
-      - clusterrole.yaml
-      - clusterrolebinding.yaml
-      - external-service.yaml
-  - name: configuration
-    kind: Apply
+        - mirror-maker-config.yaml
+        - mirror-maker.yaml
+  - name: user-workload
+    kind: Toggle
     spec:
+      parameter: RUN_USER_WORKLOAD
       resources:
-      - serviceaccount.yaml
-      - rolebinding.yaml
-      - role.yaml
-      - jaas-config.yaml
-      - krb5-config.yaml
-      - server.properties.yaml
-      - bootstrap.yaml
-      - metrics-config.yaml
-      - health-check.yaml
-      - enable-tls.yaml
+        - user-workload.yaml
+  # task used to restrict patching sts where it is not allowed
   - name: not-allowed
     kind: Dummy
```

Comment:
Dummy tasks do not have resources:
```
  - name: error
    kind: Dummy
    spec:
      wantErr: true
      fatal: false
      done: true
```
what you want here is probably `wantErr: true` and `fatal: true`?

Reply:
yeah plan should fail ideally, let me test with that here

Reply:
Nice, didn't know that's an option. Even better, but as I've said above, the correct way would be immutable parameters.

Reply:
@zen-dog this works but later the plan is stuck forever in that `FATAL_ERROR` and no way to recover from that.
```
Plan not-allowed (serial strategy) [FATAL_ERROR], last updated 2020-03-19 13:32:23
    │   └── Phase not-allowed (serial strategy) [FATAL_ERROR]
    │       └── Step not-allowed [FATAL_ERROR] (Error during execution: fatal error: default/kafka-instance failed in not-allowed.not-allowed.not-allowed.not-allowed: dummy error)
```
 I think we want to avoid/prohibit users to update certain parameters. And in case they update to let them know that it's not possible and still continue with other plans later. Right now this `not-allowed` is triggered and finished

Reply:
In that case, you could simply use `done: true` (and remove the error fields). Here a dummy task will end immediately (and successfully).

Reply:
added and works as expected!